providers/fips/self_test_kats.c

```diff
@@ -647,6 +647,206 @@ static int self_test_asym_cipher(const ST_KAT_ASYM_CIPHER *t, OSSL_SELF_TEST *st
+    size_t secretlen;
+
+    OSSL_SELF_TEST_onbegin(st, OSSL_SELF_TEST_TYPE_KAT_KEM,
+                           reject ? OSSL_SELF_TEST_DESC_DECAP_KEM_FAIL
+                                  : OSSL_SELF_TEST_DESC_DECAP_KEM);
+
+    if (reject) {
+        cipher_text = alloced = OPENSSL_memdup(t->cipher_text,
+                                               t->cipher_text_len);
+        if (alloced == NULL)
+            goto err;
+        /* Corrupt the first byte to ensure a failure */
+        *alloced ^= 1;
+    }
+
+    ctx = EVP_PKEY_CTX_new_from_pkey(libctx, pkey, "");
+    if (ctx == NULL)
+        goto err;
+
+    if (EVP_PKEY_decapsulate_init(ctx, NULL) <= 0)
+        goto err;
+
+    /* Get output size */
+    if (EVP_PKEY_decapsulate(ctx, NULL, &secretlen,
+                             cipher_text, t->cipher_text_len) <= 0)
+        goto err;
+    secret = OPENSSL_malloc(secretlen);
+    if (secret == NULL)
+        goto err;
+
+    /* Decapsulate */
+    if (EVP_PKEY_decapsulate(ctx, secret, &secretlen,
+                             cipher_text, t->cipher_text_len) <= 0)
+        goto err;
+
+    /* Compare output */
+    OSSL_SELF_TEST_oncorrupt_byte(st, secret);
+    if (secretlen != t->secret_len
+            || (memcmp(secret, t->secret, t->secret_len) != 0) == !reject)
+        goto err;
+
+    ret = 1;
+ err:
+    OPENSSL_free(alloced);
+    OPENSSL_free(secret);
+    EVP_PKEY_CTX_free(ctx);
+    OSSL_SELF_TEST_onend(st, ret);
+    return ret;
+}
+
+/*
+ * FIPS 140-3 IG 10.3.A resolution 14 mandates a CAST for ML-KEM
+ * key generation.
+ */
+static EVP_PKEY *self_test_kem_keygen(const ST_KAT_KEM *t, OSSL_SELF_TEST *st,
+                                      OSSL_LIB_CTX *libctx)
+{
+    EVP_PKEY_CTX *genctx;
+    EVP_PKEY *ret = NULL, *r = NULL;
+    OSSL_PARAM params[2] = { OSSL_PARAM_END, OSSL_PARAM_END };
+    unsigned char *buf = NULL;
+    const size_t s = t->public_key_len < t->private_key_len ? t->private_key_len
+                                                            : t->public_key_len;
+
+    OSSL_SELF_TEST_onbegin(st, OSSL_SELF_TEST_TYPE_KAT_KEM,
+                           OSSL_SELF_TEST_DESC_KEYGEN_KEM);
+
+    genctx = EVP_PKEY_CTX_new_from_name(libctx, t->algorithm, "");
+    if (genctx == NULL || EVP_PKEY_keygen_init(genctx) <= 0)
+        goto err;
+    *params = OSSL_PARAM_construct_octet_string(OSSL_PKEY_PARAM_ML_KEM_SEED,
+                                                (unsigned char *)t->seed,
+                                                t->seed_len);
+    if (!EVP_PKEY_CTX_set_params(genctx, params)
+            || !EVP_PKEY_keygen(genctx, &r))
+        goto err;
+
+    /* Allocate output space */
+    buf = OPENSSL_malloc(s);
+    if (buf == NULL)
+        goto err;
+
+    /* Compare outputs */
+    *params = OSSL_PARAM_construct_octet_string(OSSL_PKEY_PARAM_PRIV_KEY,
+                                                buf, s);
+    if (!EVP_PKEY_get_params(r, params))
+        goto err;
```

Comment:
Once the encoders PR goes in, this will not return the output you're looking for. You'll need to ask for `OSSL_PARAM_ENCODED_PRIVATE_KEY` to get the NIST expanded form, rather than just input seed back.